### PR TITLE
[FEATURE] Make some internal data structures simd aware.

### PR DESCRIPTION
**Pull request by rrahn**
* Adapts the alignment optimum to work for vectorised alignments.
* Adapts the alignment algorithm state to work for vectorised alignments.

### CODE REVIEW

File: include/seqan3/alignment/matrix/alignment_optimum.hpp [hunk 1]
```diff
@@ -14,74 +14,118 @@
...
  *
- * This function object is used in std::max functions to compare two seqan3::detail::alignment_optimum objects.
+ * Stores the optimal score of the alignment computation and the corresponding indices of the cell with the optimal
+ * score within the alignment matrix.
+ * In case the optimum is used for the vectorised alignment computation this optimum stores the optimal scores and the
+ * respective cells as simd vectors.
  */
-struct alignment_optimum_compare_less
+template <typename score_t>
+struct alignment_optimum
```
Comment (marehr) [inline include/seqan3/alignment/matrix/alignment_optimum.hpp]:
```suggestion
struct alignment_optimum
#if SEQAN3_DOXYGEN_ONLY(1)0
{}
#endif
; 
```

File: include/seqan3/alignment/matrix/alignment_optimum.hpp [hunk 2]
```diff
@@ -14,74 +14,118 @@
...
- * construction.
+ * \tparam index_t The type of the indices; must model std::unsigned_integral or seqan3::simd_concept.
  */
-template <arithmetic score_t>
-struct alignment_optimum
+template <typename index_t>
+//!\cond
+    requires std::unsigned_integral<index_t> || simd_concept<index_t>
+//!\endcond
+struct alignment_optimum_coordinate
```
Comment (marehr) [inline include/seqan3/alignment/matrix/alignment_optimum.hpp]: merge this with below

File: include/seqan3/alignment/matrix/alignment_optimum.hpp [hunk 3]
```diff
@@ -14,74 +14,118 @@
...
 {
-    //!\brief The optimal score.
-    score_t score{std::numeric_limits<score_t>::lowest()};
-    //!\brief The corresponding coordinate within the alignment matrix.
-    alignment_coordinate coordinate{};
+    index_t col_index{}; //!< The column index in the alignment matrix.
```
Comment (marehr) [inline include/seqan3/alignment/matrix/alignment_optimum.hpp]:
```suggestion
    index_t column_index{}; //!< The column index in the alignment matrix.
```

File: include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp [hunk 4]
```diff
@@ -162,18 +153,30 @@ class find_optimum_policy
...
+
+    /*!\brief Tests if the score in the current cell is greater than the current alignment optimum.
+     * \tparam cell_t The type of the alignment matrix cell.
+     * \tparam score_t The alignment algorithm score type.
+     *
+     * \param[in] cell The current cell to get the score and the coordinate from.
+     * \param[in,out] state The state with the current optimum to update.
+     */
+    template <typename cell_t, typename score_t>
+    constexpr void check_and_update(cell_t const & cell, alignment_algorithm_cache<score_t> & state) const noexcept
```
Comment (marehr) [inline include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp]:
```suggestion
    constexpr void check_and_update(cell_t const & cell, alignment_algorithm_state<score_t> & state) const noexcept
```
Don't forget this after rebase.

File: test/unit/alignment/matrix/alignment_optimum_test.cpp [hunk 5]
```diff
@@ -10,66 +10,134 @@
...
+#include <seqan3/core/simd/concept.hpp>
+#include <seqan3/core/type_traits/lazy.hpp>
+#include <seqan3/core/type_traits/template_inspection.hpp>
+
+#include <seqan3/test/simd_utility.hpp>
 
 using namespace seqan3;
 
-TEST(alignment_optimum, construction)
+template <typename not_simd_type>
```
Comment (marehr) [inline test/unit/alignment/matrix/alignment_optimum_test.cpp]:
```suggestion
template <typename scalar_type>
```

File: test/unit/alignment/matrix/alignment_optimum_test.cpp [hunk 6]
```diff
@@ -10,66 +10,134 @@
...
+    using type = typename simd_traits_t::scalar_type;
+};
+
+template <typename test_t>
+struct alignment_optimum_test : public ::testing::Test
+{
+    using scalar_lazy_t = detail::lazy_conditional_t<simd_concept<test_t>,
+                                                     detail::lazy<simd_traits, test_t>,
+                                                     test_t>;
+    using scalar_t = typename extract_scalar_type<scalar_lazy_t>::type;
```
Comment (marehr) [inline test/unit/alignment/matrix/alignment_optimum_test.cpp]: simplify me :) Use TypeParam directly.

File: include/seqan3/alignment/matrix/alignment_optimum.hpp [hunk 7]
```diff
@@ -14,74 +14,119 @@
...
- * construction.
+ * Stores the optimal score of the alignment computation and the corresponding indices of the cell with the optimal
+ * score within the alignment matrix.
+ * In case the optimum is used for the vectorised alignment computation this optimum stores the optimal scores and the
+ * respective cells as simd vectors.
  */
-template <arithmetic score_t>
+template <typename score_t>
 struct alignment_optimum
+#if SEQAN3_DOXYGEN_ONLY(1)0
```
Comment (smehringer) [inline include/seqan3/alignment/matrix/alignment_optimum.hpp]: Whats this hacky thing doing?

File: include/seqan3/alignment/matrix/alignment_optimum.hpp [hunk 8]
```diff
@@ -14,74 +14,119 @@
...
 {
-    //!\brief The optimal score.
+    //!\brief The index type used to store the alignment coordinates of the optimum.
+    using index_t = IMPLEMENTATION_DEFINED;
```
Comment (smehringer) [inline include/seqan3/alignment/matrix/alignment_optimum.hpp]: Whatis this IMPLEMENTATION_DEFINED?

File: include/seqan3/alignment/matrix/alignment_optimum.hpp [hunk 9]
```diff
@@ -14,74 +14,119 @@
...
+struct alignment_optimum<score_t>
+{
+    using scalar_t = typename simd_traits<score_t>::scalar_type;
+
+    score_t column_index{};
+    score_t row_index{};
+    score_t score{simd::fill<score_t>(std::numeric_limits<scalar_t>::lowest())};
+
+    template <std::integral column_index_t, std::integral row_index_t>
+    constexpr void update_if_new_optimal_score(score_t const & compare_score,
```
Comment (smehringer) [inline include/seqan3/alignment/matrix/alignment_optimum.hpp]: why constexpr? Will this ever be evaluated in a constexpr context? (And please don't tell me you are preparing compile time alignments :smile: )

File: include/seqan3/alignment/pairwise/detail/alignment_algorithm_state.hpp [hunk 10]
```diff
@@ -29,21 +31,22 @@ namespace seqan3::detail
  * the underlying two-dimensional matrix.
  */
 template <typename score_type>
+//!\cond
+    requires arithmetic<score_type> || simd_concept<score_type>
```
Comment (smehringer) [inline include/seqan3/alignment/pairwise/detail/alignment_algorithm_state.hpp]: also adapt the documentation

File: include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp [hunk 11]
```diff
@@ -173,20 +173,14 @@ class find_optimum_policy
      *
      * \param[in] cell The current cell to get the score and the coordinate from.
      * \param[in,out] state The state with the current optimum to update.
-     *
-     * \details
-     *
-     * Checks for a new cell in the alignment matrix. If the given score of the alignment matrix cell is greater than
```
Comment (smehringer) [inline include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp]: why did you delete this? 

EDIT: Do you need this function at all now, since you are only calling another function?

File: include/seqan3/alignment/matrix/alignment_optimum.hpp (hunk 7 above)
Comment (eseiler) [reply include/seqan3/alignment/matrix/alignment_optimum.hpp]: If it's building the docs, it uses the code below, otherwise not (then it just declares alignment_optimum).
I guess that's because doxygen can't handle the overloads for simd and arithmetic and will complain about not documenting, documenting twice or whatever.

Comment (smehringer) [reply include/seqan3/alignment/matrix/alignment_optimum.hpp]: ok. then I also understand the IMPLEMENTATION_DEFINED nothingness :smile:

File: include/seqan3/alignment/matrix/alignment_optimum.hpp (hunk 8 above)
Comment (rrahn) [reply include/seqan3/alignment/matrix/alignment_optimum.hpp]: it is an implementation defined type. For arithmetic score types it is size_t and for simd types it is the simd type. But you don't care really, you just know that it depends on the implementation.

File: include/seqan3/alignment/matrix/alignment_optimum.hpp (hunk 9 above)
Comment (rrahn) [reply include/seqan3/alignment/matrix/alignment_optimum.hpp]: > And please don't tell me you are preparing compile time alignments 😄 

Of course I thought about it because there are constexpr vector types in the future. 🤓 
But irrespective of that, all functions in the alignments are constexpr (or at least ought to be). It doesn't hurt and there is no reason not to do it.

File: include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp (hunk 11 above)
Comment (rrahn) [reply include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp]: it does reduce the structured binding in all the other functions. Makes the documentation cleaner and also the code more readable. So I don't mind keeping it.
